### DISCUSSION TITLE
Why am I not getting all the full ring?

### DISCUSSION THREAD

**Discussion by tansanDOTeth** [flauwekeul/honeycomb#93] [Q&A] [answered]
I am testing this code, and I am not sure why this isn't returning back the expected hexagons.

Code:
```
import { Grid, defineHex, rectangle, ring, spiral } from 'honeycomb-grid'

const Tile = defineHex({ dimensions: 30, orientation: "FLAT" })
const grid = new Grid(Tile, rectangle({ width: 1000, height: 1000 }))
const traverseStrategy = ring({ center: [0, 0], radius: 2 })
grid.traverse(traverseStrategy).forEach(console.log)
```

Output:
```
-------------------------------------------------
$ node index.js
$ { q: 2, r: 0 }
$ { q: 1, r: 1 }
$ { q: 0, r: 2 }
$ { q: 2, r: -1 }
```

1. I would expect 19 hexagons for "radius" 2. What am I doing wrong in the above?
1. Could someone explain what `dimensions` mean in `defineHex`? is this the number of sides? I tried Googling "dimensions" in the context of a hexagon and I don't see any definition that might explain what that means. I see it creates an xRadius, and yRadius, but unsure what this is for. A visual illustration would be much appreciated!
1. I have a similar issue with spiral. It seems to be outputting incorrect hex coordinates. Am I doing something wrong? Code below.

Code:
```
import { Grid, defineHex, rectangle, ring, spiral } from 'honeycomb-grid'

const Tile = defineHex({ dimensions: 30, orientation: "FLAT" })
const grid = new Grid(Tile, rectangle({ width: 1000, height: 1000 }))

const traverseStrategy = spiral({ start: [0, 0], radius: 5 })
grid.traverse(traverseStrategy).forEach(console.log)
```

Output:
```
$ node index.js
$ { q: 0, r: 0 }
$ { q: 1, r: 0 }
$ { q: 0, r: 1 }
**$ { q: 2, r: -1 }**
$ { q: 2, r: 0 }
$ { q: 1, r: 1 }
$ { q: 0, r: 2 }
$ { q: 3, r: -1 }
$ { q: 3, r: 0 }
$ { q: 2, r: 1 }
$ { q: 1, r: 2 }
$ { q: 0, r: 3 }
$ { q: 4, r: -2 }
$ { q: 4, r: -1 }
$ { q: 4, r: 0 }
$ { q: 3, r: 1 }
$ { q: 2, r: 2 }
$ { q: 1, r: 3 }
$ { q: 0, r: 4 }
$ { q: 5, r: -2 }
$ { q: 5, r: -1 }
$ { q: 5, r: 0 }
$ { q: 4, r: 1 }
$ { q: 3, r: 2 }
$ { q: 2, r: 3 }
$ { q: 1, r: 4 }
$ { q: 0, r: 5 }
````

The fourth coordinate seems out of place. I would have expected `q: -1, r: 1`.

![image](https://user-images.githubusercontent.com/92072158/210206145-324bef40-da51-474b-9222-7d636fedbf0e.png)

**Comment by flauwekeul** [comment, marked as answer]
Hi,

1. https://abbekeultjes.nl/honeycomb/guide/traversing-grids.html please read the documentation. Coordinates `[0, 0]` are in the top left corner of the rectangle. When you traverse outside this rectangle, by default, any hexes outside are skipped.
2. https://abbekeultjes.nl/honeycomb/api/interfaces/HexOptions.html#dimensions
3. See 1

**Reply by tansanDOTeth** [reply]
1. Ah that makes a lot of sense. Is there a way to generate a spiral by X hexagons? Ex, I know I need X hexagons, but would like it in the shape of a spiral. I am guessing I would probably have to generate the rectangle first, but that would be pretty wasteful.
2. I came across that and even tried looking at the code, but it didn't have clarify the necessity of that. Is there anything else I can read?
3. 👍

Thank you for the reply! I appreciate the help.

**Reply by flauwekeul** [reply]
1. Not really. You don't need to start with a rectangle, you can create a grid with any traverser:
    ```ts
    const Tile = defineHex({ dimensions: 30, orientation: Orientation.FLAT })
    const grid = new Grid(Tile, spiral({ start: [0, 0], radius: 2 })).forEach(console.log)
    ```
    How come you know the amount of hexes the spiral should contain? I would expect you know the radius (or rings of the spiral).
2. Dimensions is explained visually [here](https://abbekeultjes.nl/honeycomb/guide/custom-hexes.html#using-definehex). I hope this makes things clear.

**Reply by tansanDOTeth** [reply]
1. Thanks! I'll try that in the morning tomorrow. I want to draw a hexagon for each datapoint I have, which is the reason for asking if you had a suggestion for how to achieve that. Right now, it seems like I would have to pre-calculate all the rings and then find the appropriate ring level before creating my grid. Do you have a better suggestion for achieving this?
2. Thank you for linking that visual! That makes a lot of sense now. I was really confused how the hexagon could have the same x and y radius, but that visual explains it.

**Reply by tansanDOTeth** [reply]
@flauwekeul Looks like that worked! Thank you! The only unexpected thing is how it jumps to a hexagon so it's not really a spiral, but there is a fix here: https://github.com/flauwekeul/honeycomb/issues/60

Edit: Actually, that looks out of date now. Is there a way to change the spiral in the new API? https://github.com/flauwekeul/honeycomb/blob/master/src/grid/traversers/spiral.ts

![image](https://user-images.githubusercontent.com/92072158/210531467-2e866be2-278c-4f52-83c7-2cd31986fe2b.png)

**Reply by flauwekeul** [reply]
Currently, `spiral()` can't be configured to do what you need. I'll look into a solution. I'll get back to you within a week or so.

**Reply by tansanDOTeth** [reply]
> Currently, `spiral()` can't be configured to do what you need. I'll look into a solution. I'll get back to you within a week or so.

I ended up writing my own algorithm to keep traversing until X hex is created! Thanks for the library, really appreciate the design to let us traverse.

**Reply by flauwekeul** [reply]
Cool. Did you create a custom traverser?

**Reply by tansanDOTeth** [reply]
> Cool. Did you create a custom traverser?

Yeah, that's what I did in the end!

**Comment by flauwekeul** [comment]
Tomorrow I have more time, I'll answer your questions then.

Op ma 2 jan. 2023 11:26 schreef tansan.eth ***@***.***>:

**Reply by tansanDOTeth** [reply]
🙇
